### PR TITLE
Implement #266

### PR DESCRIPTION
**Pull request by BigRoy**
- Instances have a unique id generated with uuid.uuid4()

### CODE REVIEW

**Comment by BigRoy** [comment]
This will likely need some changes in `pyblish-qml` to work correctly.

File: tests/test_context.py [hunk 1]
```diff
@@ -56,19 +56,19 @@ def test_instance_equality():
...
-    context.create_instance("MyInstanceA")
-    context.create_instance("MyInstanceB")
+    instanceA = context.create_instance("MyInstanceA")
+    instanceB = context.create_instance("MyInstanceB")
 
-    assert context["MyInstanceA"].name == "MyInstanceA"
-    assert context["MyInstanceB"].name == "MyInstanceB"
-    assert context.get("MyInstanceA").name == "MyInstanceA"
-    assert context.get("MyInstanceB").name == "MyInstanceB"
+    assert context[instanceA.id].name == "MyInstanceA"
```
Comment (mottosso) [inline tests/test_context.py:68]: This is nice!

**Comment by mottosso** [comment]
Make sure you get the str-representation of the uuid, we'll need it plain and serialisable.

``` python
>>> import pyblish.api
>>> context = pyblish.api.Context()
>>> instance = context.create_instance("MyInstance")
>>> instance.id
UUID('17840082-b0f6-4eab-8318-34720ff8a3f7')  # <- Should be str
```

**Comment by BigRoy** [comment]
> Make sure you get the str-representation of the uuid, we'll need it plain and serialisable.

Thanks, totally missed that. Changed that with https://github.com/pyblish/pyblish-base/pull/267/commits/af70d41c9cc3d43d7e9e88c4a9c081016768dcbd

**Comment by mottosso** [comment]
+0.004% coverage increase, I like it. :)

**Comment by mottosso** [comment]
Great work, Roy. Having a closer look now.

**Comment by mottosso** [comment]
Ok, not sure we're pushing our luck, or whether we should go for consistency right away, but what do you think about adding `id` to Plugins and Context as well?

**Comment by BigRoy** [comment]
> Ok, not sure we're pushing our luck, or whether we should go for consistency right away, but what do you think about adding id to Plugins and Context as well?

Yes, exactly. Problem I have is figuring out where and when to add those since Plugins aren't always initialized and as such when does the uuid get created? Upon registering?

For `Context` I think it's the same as with `Instance` that it does get initialized and could just be added to `__init__`?

**Comment by mottosso** [comment]
> For Context I think it's the same as with Instance that it does get initialized and could just be added to **init**?

I think so too.

> Upon registering?

Could do, or rather in the discover() function. An alternative is the [MetaPlugin](https://github.com/pyblish/pyblish-base/blob/62d767f4386d33e7cb18f544086ac128af02c1bc/pyblish/plugin.py#L174).

discover() is pretty clean, but there are cases where one might call a plug-in directly, or pass it directly to another function, and in those cases it would not be getting an id. :(

The MetaPlugin runs on module import and would be more bulletproof.

**Comment by BigRoy** [comment]
MetaPlugin could work fine I guess.

Not sure how that works with PRs, but can you actually take over from here? (Have to run off and busy for a day or so)

**Comment by mottosso** [comment]
> Can you take over from here?

Can do!

**Comment by mottosso** [comment]
Continued in #264

**Pull request closed without merging**